pkg/bitwarden/cli.go

```diff
@@ -0,0 +1,161 @@
+package bitwarden
+
+import (
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+)
+
+type bitwardenCliClient struct {
+	run func(cmd string, args ...string) ([]byte, error)
+	sd  *sessionData
+}
+
+func newBitwardenCliClient() Client {
+	return bitwardenCliClient{
+		run: defaultCliRun,
+		sd:  &sessionData{},
+	}
+}
+
+type bwLoginResponse struct {
+	Success bool `json:"success"`
+	Data    struct {
+		Raw string `json:"raw"`
+	} `json:"data"`
+}
+
+var (
+	defaultCliRun = run
+)
+
+type sessionData struct {
+	sync.Mutex
+	session    string
+	savedItems []Item
+}
+
+func (c bitwardenCliClient) Login(username, password string) error {
+	c.sd.Lock()
+	defer c.sd.Unlock()
+	output, err := c.run("bw", "login", username, password, "--response")
+	if err != nil {
+		return err
+	}
+	r := bwLoginResponse{}
+	if err := json.Unmarshal(output, &r); err != nil {
+		//TODO need to censor if we plan this on a job
+		return err
+	}
+	if r.Success {
+		raw := r.Data.Raw
+		if raw != "" {
+			c.sd.session = raw
+			return nil
+		}
+		// should never happen
+		return fmt.Errorf("bw-login succeeded with empty '.data.raw'")
+	}
+	// should never happen
+	return fmt.Errorf("bw-login failed without run's error")
+}
+
+func (c bitwardenCliClient) listItems() ([]Item, error) {
+	var items []Item
+	if c.sd.session == "" {
+		return items, fmt.Errorf("not yet login to Bitwarden")
+	}
+	if len(c.sd.savedItems) > 0 {
+		return c.sd.savedItems, nil
+	}
+	out, err := c.run("bw", "--session", c.sd.session, "list", "items")
+	if err != nil {
+		return items, err
+	}
+	err = json.Unmarshal(out, &items)
+	if err != nil {
+		return items, err
+	}
+	return items, nil
+}
+
+func (c bitwardenCliClient) GetField(itemName, filedName string) (string, error) {
+	if c.sd.session == "" {
+		return "", fmt.Errorf("not yet login to Bitwarden")
+	}
+	if len(c.sd.savedItems) == 0 {
```

Comment:
make `listItems()` encapsulated and opaque -- a caller should never need to know that `listItems()` has caching behavior